lib/charms/mysql/v0/async_replication.py

```diff
@@ -0,0 +1,524 @@
+                )
+                event.defer()
+                return
+            logger.debug("Syncing credentials from primary cluster")
+            self._charm.unit.status = MaintenanceStatus("Syncing credentials")
+
+            credentials = self._async_replication_credentials()
+            sync_keys = {
+                SERVER_CONFIG_PASSWORD_KEY: SERVER_CONFIG_USERNAME,
+                CLUSTER_ADMIN_PASSWORD_KEY: CLUSTER_ADMIN_USERNAME,
+            }
+
+            for key, password in credentials.items():
+                # sync credentials only for necessary users
+                if key not in sync_keys:
+                    continue
+                self._charm._mysql.update_user_password(sync_keys[key], password)
+                self._charm.set_secret("app", key, password)
+                logger.debug(f"Synced {sync_keys[key]} password")
+
+            self._charm.unit.status = MaintenanceStatus("Dissolving replica cluster")
+            self._charm._mysql.dissolve_cluster()
+            # reset the cluster node count flag
+            del self._charm.app_peer_data["units-added-to-cluster"]
+
+            self._charm.unit.status = MaintenanceStatus("Populate endpoint")
+
+            # this cluster name is used by the primary cluster to identify the replica cluster
+            self.relation_data["cluster-name"] = self._charm.app_peer_data["cluster-name"]
+            # the reachable endpoint address
+            self.relation_data["endpoint"] = self._get_endpoint()
+            # the node label in the replica cluster to be created
+            self.relation_data["node-label"] = self._charm.unit_label
+
+            logger.debug("Data for adding replica cluster shared with primary cluster")
+
+            self._charm.unit.status = WaitingStatus("Waiting for primary cluster")
+        elif state == States.READY:
+            # update status
+            logger.debug("Replica cluster primary is ready")
+
+            # sync cluster-set domain name across clusters
+            if cluster_set_domain_name := self._charm._mysql.get_cluster_set_name():
+                self._charm.app_peer_data["cluster-set-domain-name"] = cluster_set_domain_name
+
+            self._charm._on_update_status(None)
+        elif state == States.RECOVERING:
+            # recoveryng cluster (copying data and/or joining units)
+            self._charm.app.status = MaintenanceStatus("Recovering replica cluster")
+            self._charm.unit.status = WaitingStatus("Waiting for recovery to complete")
+            logger.debug("Awaiting other units to join the cluster")
+            # reset the number of units added to the cluster
+            # this will trigger secondaries to join the cluster
+            node_count = self._charm._mysql.get_cluster_node_count()
+            self._charm.app_peer_data["units-added-to-cluster"] = str(node_count)
+            event.defer()
+
```

Comment:
true, but I'll keep it bc this is just a cosmetic aid, e.g. it will make the state transition from RECOVERING to READY, and consequently the unit status from `waiting` to `active` faster.
If the leadership changes, the drawback will only be that the unit will change status on the next update status.